File: projects/08-bloomFilter/implementation/bloomFilter.cpp
#include <cmath>
#include <iostream>
#include <set>
#include <vector>

uint64_t FNV_prime = 0x100000001b3;
// This is a custom hash function, not part of the Bloom Filter itself
// Typically the seed is 0xcbf29ce484222325, but we can supply custom seeds
uint64_t fnv1a_hash(std::string data, uint64_t seed) {
    // Standard FNV-1a hash algorithm. Takes the seed and XORs with each character
    // then multiplies with the FNV prime
    uint64_t hash = seed;
    for (char c : data) {
        hash = hash ^ c;
        hash = hash * FNV_prime;
    }

    return hash;
}

class BloomFilter {
    std::vector<bool> bit_array;
    int hash_functions;
    uint64_t seed1 = 0xcbf29ce484222325; // standard FNV-1a seed
    uint64_t seed2 = 0xdeadbeef; // random stuff lol

public:
    BloomFilter(int bits, int hash_functions)
        : bit_array(bits, false),
        hash_functions(hash_functions) {
    }

    void add(const std::string& str) {
        // Our two hash functions (same algo with different seeds)
        uint64_t hash1 = fnv1a_hash(str, seed1);
        uint64_t hash2 = fnv1a_hash(str, seed2);

        // Double hashing to generate more hash functions
        // For each function compute its hash, and set the corresponding bit in the bit array
        for (uint64_t i = 0; i < hash_functions; i++) {
            uint64_t index = (hash1 + i * hash2) % bit_array.size();
            bit_array[index] = true;
        }
    }

    // Checks if an element is possibly in the set
    bool contains(const std::string& element) const {
        uint64_t hash1 = fnv1a_hash(element, seed1);
        uint64_t hash2 = fnv1a_hash(element, seed2);

        // Double hashing again
        // For each function compute its hash, check if the corresponding bit is set
        // If any bit is not set, the element is definitely not in the set
        for (uint32_t i = 0; i < hash_functions; ++i) {
            size_t index = (hash1 + i * hash2) % bit_array.size();
            if (!bit_array[index]) {
                return false;
            }
        }
        return true; // could be a false positive though
    }


};

// constant used later
double ln2pow2 = std::pow(std::log(2), 2);

int main() {
    int testCases;
    std::cin >> testCases;

    for (int i = 0; i < testCases; i++) {
        // input
        int expectedItems, entries, queries;
        double falsePosRate;
        std::cin >> expectedItems >> falsePosRate >> entries >> queries;

        // determine size of bloom filter (in bits) = -(expectedItems * ln(falsePosRate)) / (ln(2)^2)
        // matches python implementation
        int bloomSize = static_cast<int>(-(expectedItems * std::log(falsePosRate)) / ln2pow2);
        bloomSize = std::max(1, bloomSize);

        // determine amount of hash functions to use
        int numHashFunctions;
        if (entries <= 0) {
            numHashFunctions = 1;
        } else {
            // hash function count (k) = (bloomSize/expectedItems) * ln(2)
            // matches python implementation
            numHashFunctions = static_cast<double>(bloomSize) / expectedItems * std::log(2);
            numHashFunctions = std::max(1, numHashFunctions);
        }

        std::cout << "m=" << bloomSize << " k=" << numHashFunctions <<
            " n=" << expectedItems << " p=" << falsePosRate << "\n";

        // let's do it! keep a ground truth set as well so we can see false positives
        BloomFilter filter = BloomFilter(bloomSize, numHashFunctions);
        std::set<std::string> groundTruth;

        // adding entries
        for (int j = 0; j < entries; j++) {
            std::string entry;
            std::cin >> entry;
            filter.add(entry);
            groundTruth.insert(entry);
        }

        // our test.txt has a --- separator
        std::string separator;
        std::cin >> separator;
        if (separator != "---") {
            std::cerr << "Could not find the --- separator\n";
            return 1;
        }

        // query the filter
        for (int j = 0; j < queries; j++) {
            std::string query;
            std::cin >> query;
            if (filter.contains(query) && groundTruth.contains(query)) {
                std::cout << query << " member\n";
            } else if (!filter.contains(query)) {
                std::cout << query << " absent\n";
            } else if (filter.contains(query) && !groundTruth.contains(query)) {
                std::cout << query << " false_positive\n";
            }
        }

        std::cout << "\n";
    }

    // yay we're done

    return 0;
}

File: projects/08-bloomFilter/implementation/BloomFilter.py
# Run: python3 BloomFilter.py io/test_1.txt
#   or redirect stdin: python3 BloomFilter.py < io/test_1.txt
# Deps: pip install -r requirements.txt (same folder)

import math
import sys

import mmh3
from bitarray import bitarray


class BloomFilter:
    """
    Bloom filter: one bit array, several hash positions per key.
    Uses mmh3 (MurmurHash3-style 32-bit hashes) with different seeds as the hash functions.
    """

    def __init__(self, items_count, fp_prob):
        """
        items_count : int
            Expected number of distinct inserts for sizing (n in the usual formulas).
        fp_prob : float
            Target false-positive probability p (0 < p < 1) used to choose bit length and k.
        """
        # Desired false-positive rate (design input, not recomputed from the bit array later).
        self.fp_prob = fp_prob

        # Number of bits in the filter (m).
        self.size = self.get_size(items_count, fp_prob)

        # How many hash values we take per key (k).
        self.hash_count = self.get_hash_count(self.size, items_count)

        # All bits start cleared.
        self.bit_array = bitarray(self.size)
        self.bit_array.setall(0)

    def add(self, item):
        """
        Insert a key: set bit to 1 at every index this key maps to.
        """
        for i in range(self.hash_count):
            # Seed i changes the hash output so we get k different indices for the same string.
            bit_index = mmh3.hash(item, i) % self.size
            self.bit_array[bit_index] = 1

    def check(self, item):
        """
        Lookup: same indices as add.
        Returns False if any mapped bit is still 0 (key definitely not inserted).
        Returns True if all mapped bits are 1 (may be a member or a false positive).
        """
        for i in range(self.hash_count):
            bit_index = mmh3.hash(item, i) % self.size
            if not self.bit_array[bit_index]:
                return False
        return True

    @staticmethod
    def get_size(n, p):
        """
        Bit array length m from expected item count n and false-positive probability p:

            m = -(n * ln p) / (ln 2)^2

        n : int
            Expected inserts used for design.
        p : float
            Target false-positive probability.

        Rounds down with int(); this implementation forces m >= 1 so modulo in hashing is safe.
        """
        m = -(n * math.log(p)) / (math.log(2) ** 2)
        return max(1, int(m))

    @staticmethod
    def get_hash_count(m, n):
        """
        Number of hashes k to pair with bit length m and expected count n:

            k = (m / n) * ln 2

        m : int
            Bit array length.
        n : int
            Expected inserts used for design.

        If n is 0, returns 1 so add/check still run and we avoid division by zero.
        """
        if n <= 0:
            return 1
        k = (m / n) * math.log(2)
        return max(1, int(k))


def main():
    """
    stdin or path: first line = t (number of test cases in this stream).
    Each case: line 'n p insert_count query_count',
    then insert_count strings, a line exactly '---', then query_count strings.
    Prints m, k, n, p and one line per query (member / false_positive / absent).
    """
    if len(sys.argv) > 1:
        stream = open(sys.argv[1], encoding="utf-8")
        own = True
    else:
        stream = sys.stdin
        own = False

    try:
        num_tests = int(stream.readline().strip())

        for t in range(num_tests):
            n, p, i, q = stream.readline().split()
            expected_item_count = int(n)
            false_positive_prob = float(p)
            insert_count = int(i)
            query_count = int(q)

            inserted_keys = []
            for _ in range(insert_count):
                inserted_keys.append(stream.readline().strip())

            sep = stream.readline().strip()
            if sep != "---":
                sys.exit(f"expected --- after inserts, got {sep!r}")

            query_keys = []
            for _ in range(query_count):
                query_keys.append(stream.readline().strip())

            # Ground truth for printing; the Bloom structure alone does not separate member vs FP. We use a set here.
            inserted_set = set(inserted_keys)
            bloom = BloomFilter(expected_item_count, false_positive_prob)
            for key in inserted_keys:
                bloom.add(key)

            print(
                f"m={bloom.size} k={bloom.hash_count} n={expected_item_count} p={false_positive_prob:.2f}"
            )
            for key in query_keys:
                if bloom.check(key):
                    if key in inserted_set:
                        print(f"{key}\tmember")
                    else:
                        print(f"{key}\tfalse_positive")
                else:
                    print(f"{key}\tabsent")
            if t < num_tests - 1:
                print()
    finally:
        if own:
            stream.close()


if __name__ == "__main__":
    main()
